src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -537,6 +549,66 @@ AppendShardIdToConstraintName(AlterTableCmd *command, uint64 shardId)
 
 
 /*
+ * UpdateWholeRowColumnReferencesWalker extends ColumnRef nodes that end with A_Star
+ * with the shardId contained in the given state.
+ *
+ * ColumnRefs that don't reference A_Star are not extended as catalog access isn't
+ * allowed here and we don't otherwise have enough context to disambiguate a
+ * field name that is identical to the table name.
+ */
+static bool
+UpdateWholeRowColumnReferencesWalker(Node *node, ColumnRefWalkerState *state)
+{
+	bool walkIsComplete = false;
+
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, IndexElem))
+	{
+		IndexElem *indexElem = (IndexElem *) node;
+
+		walkIsComplete = raw_expression_tree_walker(indexElem->expr,
+													UpdateWholeRowColumnReferencesWalker,
+													state);
+	}
+	else if (IsA(node, ColumnRef))
+	{
+		ColumnRef *columnRef = (ColumnRef *) node;
+		Node *lastField = llast(columnRef->fields);
+
+		if (IsA(lastField, A_Star))
+		{
+			/*
+			 * ColumnRef fields list ends with an A_Star, so we can blindly
+			 * extend the penultimate element with the shardId.
+			 */
+			Value *relnameValue;
```

Comment:
It's our style to always initialize variables when declared, which would normally mean setting `Value` to `NULL` here; however, if you declare `len` first, you could put `Value` second and immediately set it to the correct value… Also, `len` is lacking in specificity. Maybe something like this?

```c
int relnameFieldCount = list_length(columnRef->fields);
Value *relnameField = list_nth(columnRef->fields, relnameFieldCount - 2);
Assert(IsA(relnameField, String));

AppendShardIdToName(& relnameField->val.str, state->shardId);
```